### PR TITLE
Custom Editor for Adding and removing Connection providers 

### PR DESCRIPTION
**Pull request by rob1997**
image

### CODE REVIEW

**Review by kantagara — APPROVED** [review]
Very nice!

**Review by juans-chainsafe — CHANGES_REQUESTED** [review]
Getting this error after I remove and add back again the provider. The error happens when I click play in the editor:
```
ArgumentException: The Object you want to instantiate is null.
UnityEngine.Object.Instantiate (UnityEngine.Object original, UnityEngine.Transform parent, System.Boolean instantiateInWorldSpace) (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnityEngineObject.bindings.cs:382)
UnityEngine.Object.Instantiate[T] (T original, UnityEngine.Transform parent, System.Boolean worldPositionStays) (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnityEngineObject.bindings.cs:412)
UnityEngine.Object.Instantiate[T] (T original, UnityEngine.Transform parent) (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnityEngineObject.bindings.cs:407)
ChainSafe.Gaming.UnityPackage.UI.ConnectModal.Initialize (ChainSafe.Gaming.UnityPackage.Connection.ConnectionHandler connectionHandler) (at /Users/juanmanuelspoleti/Desktop/workspace/web3.unity/Packages/io.chainsafe.web3-unity/Runtime/Scripts/UI/ConnectModal.cs:43)
ConnectToWallet.Start () (at Assets/Samples/web3.unity SDK/2.6.0/Web3.Unity Samples/Scripts/Scenes/ConnectToWallet.cs:45)
System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__7_0 (System.Object state) (at <e307bbb467104258887a104f6151f183>:0)
UnityEngine.UnitySynchronizationContext+WorkRequest.Invoke () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:153)
UnityEngine.UnitySynchronizationContext.Exec () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:83)
UnityEngine.UnitySynchronizationContext.ExecuteTasks () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:107)
```

**Comment by rob1997** [comment]
> Getting this error after I remove and add back again the provider. The error happens when I click play in the editor:
> 
> ```
> ArgumentException: The Object you want to instantiate is null.
> UnityEngine.Object.Instantiate (UnityEngine.Object original, UnityEngine.Transform parent, System.Boolean instantiateInWorldSpace) (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnityEngineObject.bindings.cs:382)
> UnityEngine.Object.Instantiate[T] (T original, UnityEngine.Transform parent, System.Boolean worldPositionStays) (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnityEngineObject.bindings.cs:412)
> UnityEngine.Object.Instantiate[T] (T original, UnityEngine.Transform parent) (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnityEngineObject.bindings.cs:407)
> ChainSafe.Gaming.UnityPackage.UI.ConnectModal.Initialize (ChainSafe.Gaming.UnityPackage.Connection.ConnectionHandler connectionHandler) (at /Users/juanmanuelspoleti/Desktop/workspace/web3.unity/Packages/io.chainsafe.web3-unity/Runtime/Scripts/UI/ConnectModal.cs:43)
> ConnectToWallet.Start () (at Assets/Samples/web3.unity SDK/2.6.0/Web3.Unity Samples/Scripts/Scenes/ConnectToWallet.cs:45)
> System.Runtime.CompilerServices.AsyncMethodBuilderCore+<>c.<ThrowAsync>b__7_0 (System.Object state) (at <e307bbb467104258887a104f6151f183>:0)
> UnityEngine.UnitySynchronizationContext+WorkRequest.Invoke () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:153)
> UnityEngine.UnitySynchronizationContext.Exec () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:83)
> UnityEngine.UnitySynchronizationContext.ExecuteTasks () (at /Users/bokken/build/output/unity/unity/Runtime/Export/Scripting/UnitySynchronizationContext.cs:107)
> ```

Hey Juan, the issue you're having is actually expected because removing and adding a new provider means it's new and it loses old references, the idea for add-provider is to create the connection provider but user will still need to configure it with preferred settings and references, so until you do so the references remain empty/null so if you try and run it without setting it up it'll throw an exception.

**Review by juans-chainsafe — APPROVED** [review]
Got it, in that case is working as expected!

**Review by oleksandrchainsafe — CHANGES_REQUESTED** [review]
Left a couple of discussion points.

File: Packages/io.chainsafe.web3-unity/Editor/ConnectionHandlerEditor.cs [hunk 1]
```diff
@@ -18,23 +27,39 @@ public struct Provider
...
+            EditorGUILayout.BeginVertical();
             
+            // Get provider display name.
             var providersProperty = serializedObject.FindProperty("providers");
 
-            int arraySize = providersProperty.arraySize;
+            ConnectionProvider[] allProviders = Resources.LoadAll<ConnectionProvider>(string.Empty);
             
+            // Get available providers.
+            List<ConnectionProvider> availableProviders = new List<ConnectionProvider>();
```
[the remark this answers is not in the document]
Comment (creeppak) [reply Packages/io.chainsafe.web3-unity/Editor/ConnectionHandlerEditor.cs:59]: I know it's just an editor window, but this stuff may cause lags in playmode, so we better keep the update loop clean.

**Pull request merged**